### DISCUSSION TITLE
Asynchronous Calls using the gloo_net Request module: cannot be sent between threads safely

### DISCUSSION THREAD

**Discussion by teddywaweru** [tauri-apps/tauri#5972] [Q&A] [answered]
I'm not entirely sure of where I'm heading with this project, but I was attempting to make a GET Request through the gloo_net crate:
```
#![cfg_attr(
    all(not(debug_assertions), target_os = "windows"),
    windows_subsystem = "windows"
)]
use gloo_net::http::Request;
use serde::Deserialize;

#[derive(Clone, PartialEq, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct User {
    first_name: String,
    last_name: String,
    email: String,
    phone: String,
}

#[derive(Clone, Deserialize, PartialEq, Debug)]
pub struct Users {
    users: Vec<User>,
}
// Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
#[tauri::command]
async fn greet(name: &str) -> String {
    let users = Request::get("https://dummyjson.com/users")
        .send()
        .await
        .unwrap();
    let users = users.json::<Users>().await;
    format!("Hello, {}! You've been greeted from Rust!", name)
}

fn main() {
    tauri::Builder::default()
        .invoke_handler(tauri::generate_handler![greet])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}

```
But I keep running into an error on the future not being safe to send between threads:
![image](https://user-images.githubusercontent.com/57697272/210406529-53a37246-c536-4670-bc60-feddfaca0f40.png)
Full Error:
![image](https://user-images.githubusercontent.com/57697272/210406615-378dd000-4438-409b-8de8-055cb0075df1.png)

Is there sth I'm missing in the implementation of asynchronous request calls in the Tauri App?

**Comment by FabianLars** [comment, marked as answer]
all futures in tauri commands must be [`Send`](https://doc.rust-lang.org/nomicon/send-and-sync.html). Normally this is no problem, `!Send` futures are quite rare, and workarounds to still run them tend to be really annoying.
Buuut that shouldn't matter too much since gloo_net is a http library for WASM contexts, which is the wrong choice here. tauri commands (or everything in src-tauri/src for that matter) are not compiled to WASM but to native platform code. You'd probably be better of with crates like `reqwest`

**Reply by teddywaweru** [reply]
You were absolutely right on the http library to use: Reqwest worked as intended . 👌 Thanks a lot!
I did notice that I needed to change the &str variable passed to an owned String, otherwise I got the error below: 
Does it have sth to do with how async works on tauri?

![image](https://user-images.githubusercontent.com/57697272/210588689-a0abcf90-c86e-41ed-8124-06a7fb1b7054.png)

![image](https://user-images.githubusercontent.com/57697272/210590577-f3b20229-c47d-488a-8ddb-f172d666f925.png)

**Reply by FabianLars** [reply]
> Does it have sth to do with how async _commands_ work in tauri?

Yeah, pretty much. async commands must have a `'static` lifetime so that they can be spawned on the tokio runtime, `&str` prevents a static lifetime.
